src/text-editor.coffee

```diff
@@ -818,114 +818,126 @@ class TextEditor extends Model
 
-  # Move lines intersection the most recent selection up by one row in screen
+  # Move lines intersection the most recent selection or multiple selections up by one row in screen
   # coordinates.
   moveLineUp: ->
-    selection = @getSelectedBufferRange()
-    return if selection.start.row is 0
-    lastRow = @buffer.getLastRow()
-    return if selection.isEmpty() and selection.start.row is lastRow and @buffer.getLastLine() is ''
-
-    @transact =>
-      foldedRows = []
-      rows = [selection.start.row..selection.end.row]
-      if selection.start.row isnt selection.end.row and selection.end.column is 0
-        rows.pop() unless @isFoldedAtBufferRow(selection.end.row)
-
-      # Move line around the fold that is directly above the selection
-      precedingScreenRow = @screenPositionForBufferPosition([selection.start.row]).translate([-1])
-      precedingBufferRow = @bufferPositionForScreenPosition(precedingScreenRow).row
-      if fold = @largestFoldContainingBufferRow(precedingBufferRow)
-        insertDelta = fold.getBufferRange().getRowCount()
-      else
-        insertDelta = 1
-
-      for row in rows
-        if fold = @displayBuffer.largestFoldStartingAtBufferRow(row)
-          bufferRange = fold.getBufferRange()
-          startRow = bufferRange.start.row
-          endRow = bufferRange.end.row
-          foldedRows.push(startRow - insertDelta)
+    newSelectionBufferRanges = []
+    selections = @getSelectedBufferRanges()
+    selections.sort (a, b) ->
+      return a.compare(b)
+    for selection in selections
+      return if selection.start.row is 0
+      lastRow = @buffer.getLastRow()
+      return if selection.isEmpty() and selection.start.row is lastRow and @buffer.getLastLine() is ''
+
+      @transact =>
+        foldedRows = []
+        rows = [selection.start.row..selection.end.row]
+        if selection.start.row isnt selection.end.row and selection.end.column is 0
+          rows.pop() unless @isFoldedAtBufferRow(selection.end.row)
+
+        # Move line around the fold that is directly above the selection
+        precedingScreenRow = @screenPositionForBufferPosition([selection.start.row]).translate([-1])
+        precedingBufferRow = @bufferPositionForScreenPosition(precedingScreenRow).row
+        if fold = @largestFoldContainingBufferRow(precedingBufferRow)
+          insertDelta = fold.getBufferRange().getRowCount()
         else
-          startRow = row
-          endRow = row
+          insertDelta = 1
 
-        insertPosition = Point.fromObject([startRow - insertDelta])
-        endPosition = Point.min([endRow + 1], @buffer.getEndPosition())
-        lines = @buffer.getTextInRange([[startRow], endPosition])
-        if endPosition.row is lastRow and endPosition.column > 0 and not @buffer.lineEndingForRow(endPosition.row)
-          lines = "#{lines}\n"
+        for row in rows
+          if fold = @displayBuffer.largestFoldStartingAtBufferRow(row)
```

Comment:
I didn't really experiment with screen coordinates. I will spend some time playing around and seeing what happens